mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,800 @@
+//===- AffineLoopInterchange.cpp - Pass to perform loop interchange-----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a loop interchange pass that optimizes for locality
+// (spatial and temporal - both self and group) and parallelism for multicores,
+// to minimize the frequency of synchronization. The pass works for both
+// perfectly nested and imperfectly nested loops (any level of nesting). However
+// in the presence of affine.if statements and/or non-rectangular iteration
+// space, the pass simply bails out - leaving the original loop nest unchanged.
+// The pass is triggered by the command line flag -affine-loop-interchange.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+
+using namespace mlir;
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+};
+} // namespace
+
+/// Returns true if any affine.if op found in the loop nest rooted at `forOp`
+static bool hasAffineIfStatement(AffineForOp &forOp) {
+  auto walkResult =
+      forOp.walk([&](AffineIfOp op) { return WalkResult::interrupt(); });
+  return walkResult.wasInterrupted();
+}
+
+/// Checks if this `loopNest` has a rectangular-shaped iteration space.
+static bool isRectangularAffineForLoopNest(ArrayRef<AffineForOp> loopNest) {
+  for (AffineForOp forOp : loopNest) {
+    if (!forOp.hasConstantUpperBound() || !forOp.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
```

Comment:
This naming and API isn't really meaningful. This shouldn't really be called a matrix in the first place. Second, you could have just passed matrix[column] to this method or better just checked inline whether the SmallVector is all zero.

Reply:
Fixed. Thanks.